addons/website_event_crm/models/event_registration.py

```diff
@@ -26,6 +26,33 @@ def _get_lead_description_fields(self):
         res.append('registration_answer_ids')
         return res
 
+    def _get_lead_grouping(self, rules, rule_to_new_regs):
+        visitor_registrations = self.filtered('visitor_id')
+        # We don't want to group by visitor if there is a sale order,
+        # this is done here to avoid another bridge module.
+        if 'sale_order_id' in self._fields:
+            visitor_registrations -= visitor_registrations.filtered('sale_order_id')
+        grouping_res = super(EventRegistration, self - visitor_registrations)._get_lead_grouping(rules, rule_to_new_regs)
+
+        if visitor_registrations:
+            related_leads = self.env['crm.lead'].search([
+                ('event_lead_rule_id', 'in', rules.ids),
+                ('registration_ids', 'in', visitor_registrations.ids),
+            ])
+
+            for rule in rules:
+                rule_new_regs = rule_to_new_regs[rule]
+                visitor_to_regs = (rule_new_regs & visitor_registrations).grouped('visitor_id')
+                visitor_res = []
+                for visitor, registrations in visitor_to_regs.items():
+                    registrations = registrations.sorted('id')
+                    leads = related_leads.filtered(lambda lead: lead.event_lead_rule_id == rule and visitor in lead.registration_ids.visitor_id)
```

Comment:
While I'm here, I remember I have a hard time knowing when to use the mapped function instead of record chaining. Here we have 

`record.many2many.many2one`, so this will correctly return **all** visitors right?

But `record.many2many.many2one.many2many` would need a mapped function, is that it?

Reply:
Thanks! Hopefully I won't be confused anymore. Was basically what I had in mind :+1: